### PR TITLE
Protect: Improve Accessibility for Math Fallback.

### LINKED ISSUE #6714: Protect: Math Fallback is not accessible
When Protect's Math Fallback is displayed at the bottom of WordPress' log in form, there are no labels or indicators that you are looking at a math fallback page. This makes it difficult for folks using screen readers to actually answer the math challenge.

A potential solution is highlighted in the forum thread where this issue was reported:
https://wordpress.org/support/topic/wordpress-org-captcha-not-using-correct-structure/

### PR DESCRIPTION
**Pull request by stevenlinx**
Fixes #6714

#### Changes proposed in this Pull Request:

* Made the math question contain inside a html `<label>` element.
* Made the `<label>` tied to the `<input>` field for the answer via `for="sum"`.
* The math operators were also implemented with the plain text equivalent for people using assistive technologies since those operators are not always supported by screen readers.
* The `<input type="">` is changed to `number`.

#### Testing instructions:
1.) Checkout the code on a WP test site that you've admin access to

2.) Make sure "Protect" is enabled under WP-Admin (Jetpack -> Settings -> Security)
https://jetpack.com/support/security-features/#protect

3.) Test login with a test user. Trigger the Math captcha by attempting a fixed number of failed login using the correct username, but with the wrong password.

4.) Use a screen reader to test the reader reads out the math equation properly. Enter the right and wrong numbers to make sure the results are as intended.

5.) Once done testing, make sure the test user can login normally, then revert the test site code to its previous state.

### CODE REVIEW

**Review by oskosk — DISMISSED** [review]
LGTM! @dereksmart does it look good to you ?

File: modules/protect/math-fallback.php
Comment (eliorivero) [inline modules/protect/math-fallback.php]: Since it's a word expected to be read by screen readers, it should be made available for translation.

File: modules/protect/math-fallback.php
Comment (eliorivero) [inline modules/protect/math-fallback.php]: Same here, it should be translatable.

**Review by eliorivero — CHANGES_REQUESTED** [review]
In addition to the comments left in the code, this 
```css
style="left: -999em; position: absolute;"
```
is not bulletproof for left-to-right languages and might cause layout issues. Have you tested it with a LTR language? Maybe it's better to use top instead of left.
Also, these styles should be in a CSS file and referenced by a class so we can avoid writing inline CSS and duplication.

**Comment by stevenlin-x** [comment]
@eliorivero

1.)

I used "24 Accessibility": a blog dedicated to accessibility technology created by 3 accessibility engineers.
https://www.24a11y.com/about/

Providing Alternate Content for Screen Reader Users
https://www.24a11y.com/2017/providing-alternate-content-screen-reader-users/

The article suggested two methods:
i.) the `clip` technique
ii.) hide text off screen by outdent:

The original discussion thread used the second method (hide text off screen).

2.)

Another source CSS-Tricks.com which also recommends hide text off screen by outdent:
https://css-tricks.com/snippets/css/accessibilityseo-friendly-css-hiding/

A commenter in the same blog post recommends the `clip` technique:
https://css-tricks.com/snippets/css/accessibilityseo-friendly-css-hiding/#comment-1206657

Keith Hays:
"If your intent is to include additional information for visually impaired users, the positioning technique that Chris describes here is one of the W3 recommended techniques. Another one that I’ve seen uses clip:"

I actually liked the `clip` method better since you don't need to play with the absolute position. I've since modified the code to use `clip` method instead.

3.)

If you have other ideas, please post your solution.

4.)
I've made the rest of the changes you suggested: made the text translatable and moved the style into CSS file.

**Comment by oskosk** [comment]
@eliorivero can you check if your feedback was addressed... In that case could you dismiss your review ?

File: modules/protect/protect-dashboard-widget.css
Comment (oskosk) [inline modules/protect/protect-dashboard-widget.css]: There's an unnecessary space at the beginning of the line here

File: modules/protect/math-fallback.php
Comment (eliorivero) [inline modules/protect/math-fallback.php]: This will render `__(plus)` as text. Not sure how a screen reader would read this but it's confusing. The right syntax for the text is
```php
<?php esc_html_e( 'plus' ) ?>
```

File: modules/protect/math-fallback.php
Comment (eliorivero) [inline modules/protect/math-fallback.php]: Same here, this is won't render "equals" as text but "__(equals)". Fix it following the code in the other comment.

File: modules/protect/math-fallback.php
Comment (eliorivero) [inline modules/protect/math-fallback.php]: No `&nbsp;` please. Space this using CSS.

File: modules/protect/math-fallback.php
Comment (eliorivero) [inline modules/protect/math-fallback.php]: Again, no `&nbsp;` please.

File: modules/protect/protect-dashboard-widget.css
Comment (eliorivero) [inline modules/protect/protect-dashboard-widget.css]: This property is not correctly written.

File: modules/protect/protect-dashboard-widget.css
Comment (eliorivero) [inline modules/protect/protect-dashboard-widget.css]: We don't support IE 6/7 on front end so you can remove this line.

File: modules/protect/protect-dashboard-widget.css
Comment (eliorivero) [inline modules/protect/protect-dashboard-widget.css]: WP core uses `a11y` instead of `accessibility` when prefixing variables. It's shorter.

**Review by eliorivero — DISMISSED** [review]
Thanks for the changes. This still needs some improvement, I've left the comments on the code.
Additionally, I think those `!important` could be removed. Why did you add them? We don't have this rule anywhere else in Jetpack so it won't conflict.

**Comment by stevenlin-x** [comment]
@eliorivero 

1.)
In the above comment, you have 
`<?php esc_html_e( 'plus' ) ?>`

you missed the semicolon and the domain should be 'jetpack', same as in the 'Prove your humanity' text above? otherwise CodeSniffer would warn.
it should be:
`<?php esc_html_e( 'plus', 'jetpack' ); ?>`

2.)
I made the rest of the change you proposed.

File: modules/protect/math-fallback.php
Comment (eliorivero) [inline modules/protect/math-fallback.php]: The `esc_html` doesn't print anything. You must use `echo` here. Also, you can leave the casting aside.
```php
<?php echo esc_html( $num1 ); ?>
```

File: modules/protect/math-fallback.php
Comment (eliorivero) [inline modules/protect/math-fallback.php]: Same here, use `echo` to output this.

File: modules/protect/math-fallback.php
Comment (eliorivero) [inline modules/protect/math-fallback.php]: Same here, use `echo` to output this.

File: modules/protect/math-fallback.php
Comment (simison) [reply modules/protect/math-fallback.php]: Noting that you can also use `<?php esc_html_e(  $num1 ); ?>` to output this directly without `echo`, if you prefer.

https://codex.wordpress.org/Function_Reference/esc_attr_e
https://codex.wordpress.org/Data_Validation#Text_Nodes

Comment (stevenlin-x) [reply modules/protect/math-fallback.php]: Yes, I'm aware of `esc_html_e` and I already use it for 'plus' and 'equals' text string. I chose not to use it for `$num1` `$num2` `$ans` because they're all random numbers and retrieving the translation becomes redundant. The code responsible `$num1` `$num2` `$ans` is cut off in the code summary view.

Thanks for the help.

File: modules/protect/math-fallback.php
Comment (stevenlin-x) [reply modules/protect/math-fallback.php]: @eliorivero 
I think `esc_attr` is better here.
https://developer.wordpress.org/reference/functions/esc_attr/#comment-1162

File: modules/protect/math-fallback.php
Comment (eliorivero) [inline modules/protect/math-fallback.php]: If you look at line 120 in this file, you have
```php
$num1 = rand( 0, 10 );
$num2 = rand( 1, 10 );
...
$ans  = sha1( $salt . $sum );
```
so `esc_html` is not really needed here, or in the line outputting `$num2` or in the line outputting `$ans` as an attribute. Escaping should be performed when the input is provided by the user, which includes the case of translations, where `esc_html_e` is now properly used because translation are also a source of vulnerabilities. However, these values are generated here by code without user input so it's safe to remove the escaping.
After this is addressed, @oskosk or @simison can give a final test and approve for merge if they consider it final.

Comment (stevenlin-x) [reply modules/protect/math-fallback.php]: @eliorivero 
I initially didn't, but if you don't, the CodeSnifffer complains.

"All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbook)"
https://developer.wordpress.org/plugins/security/securing-output/#escaping

Since I already have them escaped, do you absolutely want this to be removed at this point?

Is there a second opinion on this? It's not harmful to have escaping and it also solves the warning.

**Comment by brbrr** [comment]
Hi @stevenlin-x 
Your PR is missing testing instructions. Could you please add them into PR description, so anyone would be able to easily test these changes? Thanks!

**Comment by stevenlin-x** [comment]
@brbrr

Thanks for the reminder.

The testing instructions were added a few days ago.

**Comment by brbrr** [comment]
This PR needs to be rebased with master to include the latest fix for WSOD:
```
Fatal error: Call to undefined method Jetpack_Simple_Payments::is_enabled_jetpack_simple_payments() in /srv/users/user6cc86964/apps/user6cc86964/public/wp-content/plugins/jetpack-dev/modules/widgets/simple-payments.php on line 480
```

**Comment by stevenlin-x** [comment]
@brbrr 

It's now rebased.

**Comment by brbrr** [comment]
I've followed your test instructions, and it looks like your CSS changes are not loaded in `wp-login.php` page. Please make sure that `a11y-visually-hidden` is available, and update testing instruction accordingly. Here how it looks like now:

screen shot on 2018-07-31 at 16 40 10


Moving to next release

**Pull request closed without merging**
